src/commonpp/core/json_escape.cpp

```diff
@@ -0,0 +1,192 @@
+           obtaining a copy of this software and associated documentation files
+           (the "Software"), to deal in the Software without restriction,
+           including without limitation the rights to use, copy, modify, merge,
+           publish, distribute, sublicense, and/or sell copies of the Software,
+           and to permit persons to whom the Software is furnished to do so,
+           subject to the following conditions:
+           <br>
+           The above copyright notice and this permission notice shall be
+           included in all copies or substantial portions of the Software.
+           <br>
+           THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+           EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+           MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+           NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+           BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+           ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+           CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+           SOFTWARE.
+
+@author [Niels Lohmann](http://nlohmann.me)
+@see https://github.com/nlohmann/json to download the source code
+*/
+
+#include "commonpp/core/string/json_escape.hpp"
+
+namespace commonpp
+{
+namespace string
+{
+
+/*!
+   @brief calculates the extra space to escape a JSON string
+   @param[in] s  the string to escape
+   @return the number of characters required to escape string @a s
+   @complexity Linear in the length of string @a s.
+   */
+std::size_t extra_space(const std::string& s) noexcept
+{
+    std::size_t result = 0;
+
+    for (const auto& c : s)
+    {
+        switch (c)
+        {
+        case '"':
+        case '\\':
+        case '\b':
+        case '\f':
+        case '\n':
+        case '\r':
+        case '\t':
+        {
+            // from c (1 byte) to \x (2 bytes)
+            result += 1;
+            break;
+        }
+
+        default:
+        {
+            if (c >= 0x00 and c <= 0x1f)
+            {
+                // from c (1 byte) to \uxxxx (6 bytes)
+                result += 5;
+            }
+            break;
+        }
+        }
+    }
+
+    return result;
+}
+
+/*!
+@brief escape a string
+Escape a string by replacing certain special characters by a sequence of an
+escape character (backslash) and another character and other control
+characters by a sequence of "\u" followed by a four-digit hex
+representation.
+@param[in] s  the string to escape
+@return  the escaped string
+@complexity Linear in the length of string @a s.
+*/
+std::string escape_json_string(const std::string& s) noexcept
+{
+    const auto space = extra_space(s);
+    if (space == 0)
+    {
+        return s;
+    }
+
+    // create a result string of necessary size
+    std::string result(s.size() + space, '\\');
```

Comment:
since this can fail (unlikely, very much granted, I don't really like having this as noexcept. what do you think ?

Reply:
I noticed it and though it was a bit odd as well. I don't rememer what it looks like in a core dump if noexcept throws, but maybe it's easier to debug. Most times people have some unhandle exception handler which doesn't really log where abouts in the code that the exception was raised, you'd probably just get "bad_alloc" from "somewhere"

Reply:
At best, this is unfortunately undefined behavior…
Maybe we should remove the noexcept, the licence allows it anyway :)

Reply:
yeah, I remember you trying to workout a strange crash at 360, it was a throw from within a noexcept and very confusing at the time.